contracts/javascore/xcall/src/main/java/foundation/icon/xcall/CallServiceImpl.java

```diff
@@ -519,26 +519,29 @@ private void handleResult(byte[] data) {
 
     private boolean verifyProtocols(String fromNid, String[] protocols, DictDB<String, Boolean> pendingDb) {
         Address caller = Context.getCaller();
-        if (protocols.length > 1) {
-            pendingDb.set(caller.toString(), true);
-            for (String protocol : protocols) {
-                if (!pendingDb.getOrDefault(protocol, false)) {
-                    return false;
-                }
-            }
-
-            for (String protocol : protocols) {
-                pendingDb.set(protocol, null);
-            }
+        if (protocols.length == 0) {
+            Context.require(caller.equals(defaultConnection.get(fromNid)), "ProtocolSourceMismatch");
+            return true;
         } else if (protocols.length == 1) {
             Context.require(caller.toString().equals(protocols[0]), "ProtocolSourceMismatch");
-        } else {
-            Context.require(caller.equals(defaultConnection.get(fromNid)), "ProtocolSourceMismatch");
+            return true;
         }
 
+
+        pendingDb.set(caller.toString(), true);
+        for (String protocol : protocols) {
+            if (!pendingDb.getOrDefault(protocol, false)) {
+                return false;
+            }
+        }
+
+        for (String protocol : protocols) {
+            pendingDb.set(protocol, null);
+        }
+
+
         return true;
     }
-
     private boolean isReply(String netId, String[] sources) {
```

Comment:
`
if protocols.length==0 {
protocols=[defaultConnection.get(nid)]
}
Context.require(protocols.contains(caller.toString()), "ProtocolSourceMismatch");
  pendingDb.set(caller.toString(), true);
        for (String protocol : protocols) {
            if (!pendingDb.getOrDefault(protocol, false)) {
                return false;
            }
        }

        for (String protocol : protocols) {
            pendingDb.set(protocol, null);
        }
`   .  This  also ensures only protocols can queue message . Does this makes sense @AntonAndell ?

Reply:
But this is more expensive though, and only optimized gas in invalid uses? 
I do not feel extra logic is worth the benefit in logic.

Reply:
ya this might be more expensive since we are checking validity of caller in all cases. For optimized and cleaner solution above one might be better fit. I was only  concerned about conciseness.